## Dependencies:
TensorFlow, MuJoCo, OpenAI Gym

We can use the example code run_expert.py to generate the expert trajectory data. It load up an expert policy, run a specified number of roll-outs, and save out data. The expert policies are located in /experts. 

We can perform the following to run the expert policies and store the (observation, action) pairs in /expert_data.

In [11]:
%%bash
# --render to render the video, --num_rollouts to specify the number of total trajectories tested
python run_expert.py experts/Humanoid-v2.pkl Humanoid-v2 --render --num_rollouts 40

# can try other models like: Ant-v2.pkl, HalfCheetah-v2.pkl, Hopper-v2.pkl ...

loading and building expert policy
obs (1, 376) (1, 376)
loaded and built
WARN: gym.spaces.Box autodetected dtype as <class 'numpy.float32'>. Please provide explicit dtype.
WARN: gym.spaces.Box autodetected dtype as <class 'numpy.float32'>. Please provide explicit dtype.
iter 0
Creating window glfw
100/1000
200/1000
300/1000
400/1000
500/1000
600/1000
700/1000
800/1000
900/1000
1000/1000
iter 1
100/1000
200/1000
300/1000
400/1000
500/1000
600/1000
700/1000
800/1000
900/1000
1000/1000
iter 2
100/1000
200/1000
300/1000
400/1000
500/1000
600/1000
700/1000
800/1000
900/1000
1000/1000
iter 3
100/1000
200/1000
300/1000
400/1000
500/1000
600/1000
700/1000
800/1000
900/1000
1000/1000
iter 4
100/1000
200/1000
300/1000
400/1000
500/1000
600/1000
700/1000
800/1000
900/1000
1000/1000
iter 5
100/1000
200/1000
300/1000
400/1000
500/1000
600/1000
700/1000
800/1000
900/1000
1000/1000
iter 6
100/1000
200/1000
300/1000
400/1000
500/1000
600/1000
700/1000
800/1000
900/1000
1000/1000
iter 7
100/1000
200/1

2018-10-01 22:37:29.300347: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
Instructions for updating:
Please use tf.global_variables instead.
Instructions for updating:
Use `tf.variables_initializer` instead.


We can find the specifications of these models in
https://github.com/openai/gym/wiki/Humanoid-V1
https://gym.openai.com/envs/Humanoid-v2/

Take the humanoid for example, its observation is categorized in a 376 box and its action space is about 17. We can check that just by looking at the data:

In [8]:
import pickle
with open('expert_data/Humanoid-v2.pkl', 'rb') as f:
    data = pickle.load(f)

In [25]:
print("The action data shape", data['actions'].shape)
print("The observation data shape", data['observations'].shape)

The action data shape (10000, 1, 17)
The observation data shape (10000, 376)


### Implement Behabior Cloning


In [13]:
%%bash
python run_behaviorCloning.py experts/Humanoid-v2.pkl Humanoid-v2 expert_data/Humanoid-v2.pkl --render --num_rollouts 20

loading and building expert policy
obs (1, 376) (1, 376)
loaded and built
Epoch 1/30

32000/32000 [==============================] - 1s 39us/step - loss: 0.2336 - acc: 0.2247
Epoch 2/30

32000/32000 [==============================] - 1s 29us/step - loss: 0.1530 - acc: 0.5204
Epoch 3/30

32000/32000 [==============================] - 1s 31us/step - loss: 0.1262 - acc: 0.6530
Epoch 4/30

32000/32000 [==============================] - 1s 31us/step - loss: 0.1063 - acc: 0.7127
Epoch 5/30

32000/32000 [==============================] - 1s 31us/step - loss: 0.0774 - acc: 0.7864
Epoch 6/30

32000/32000 [==============================] - 1s 29us/step - loss: 0.0573 - acc: 0.8215
Epoch 7/30

32000/32000 [==============================] - 1s 28us/step - loss: 0.0528 - acc: 0.8344
Epoch 8/30

32000/32000 [==============================] - 1s 29us/step - loss: 0.0515 - acc: 0.8403
Epoch 9/30

32000/32000 [==============================] - 1s 29us/step - loss: 0.0507 - acc: 0.8466
Epoch 10/30

3200

2018-10-01 22:43:09.565817: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
Using TensorFlow backend.
run_behaviorCloning.py:77: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.
  model.fit(X_train, Y_train, batch_size=64, nb_epoch=30, verbose=1)
Instructions for updating:
Please use tf.global_variables instead.
Instructions for updating:
Use `tf.variables_initializer` instead.
Exception ignored in: <bound method BaseSession.__del__ of <tensorflow.python.client.session.Session object at 0xd2ae38d68>>
Traceback (most recent call last):
  File "/anaconda3/envs/CS294/lib/python3.5/site-packages/tensorflow/python/client/session.py", line 686, in __del__
TypeError: 'NoneType' object is not callable


In [14]:
%%bash
python dagger.py experts/Humanoid-v2.pkl Humanoid-v2 expert_data/Humanoid-v2.pkl --render --num_rollouts 20

loading and building expert policy
obs (1, 376) (1, 376)
loaded and built
Epoch 1/30

32000/32000 [==============================] - 1s 39us/step - loss: 0.1989 - acc: 0.4720
Epoch 2/30

32000/32000 [==============================] - 1s 30us/step - loss: 0.1474 - acc: 0.6112
Epoch 3/30

32000/32000 [==============================] - 1s 30us/step - loss: 0.0952 - acc: 0.7297
Epoch 4/30

32000/32000 [==============================] - 1s 30us/step - loss: 0.0664 - acc: 0.8046
Epoch 5/30

32000/32000 [==============================] - 1s 30us/step - loss: 0.0508 - acc: 0.8213
Epoch 6/30

32000/32000 [==============================] - 1s 29us/step - loss: 0.0351 - acc: 0.8266
Epoch 7/30

32000/32000 [==============================] - 1s 29us/step - loss: 0.0242 - acc: 0.8368
Epoch 8/30

32000/32000 [==============================] - 1s 29us/step - loss: 0.0219 - acc: 0.8452
Epoch 9/30

32000/32000 [==============================] - 1s 29us/step - loss: 0.0205 - acc: 0.8474
Epoch 10/30

3200

2018-10-01 22:43:59.091956: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
Using TensorFlow backend.
dagger.py:85: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.
  model.fit(X_train, Y_train, batch_size=64, nb_epoch=30, verbose=1)
Instructions for updating:
Please use tf.global_variables instead.
Instructions for updating:
Use `tf.variables_initializer` instead.
bash: line 1: 90809 Terminated: 15          python dagger.py experts/Humanoid-v2.pkl Humanoid-v2 expert_data/Humanoid-v2.pkl --render --num_rollouts 20


(5, 35)